### PR TITLE
Make purecap libslirp available

### PR DESCRIPTION
**Pull request by markjdb**
This is useful for bhyve, which has a network backend that makes use of libslirp to provide userspace networking. Unfortunately, libslirp requires a smallish patch to work under CHERI. It works in my testing but needs more thought before it can be upstreamed; see the commit log message for some details.

Note, I only tested this by building ports locally on my morello system. I suspect that enabling glib20 for purecap will cause us to start building lots of other packages which previously had been skipped.

### CODE REVIEW

**Comment by kwitaszczyk** [comment]
@gcjenkinson Is https://github.com/CTSRD-CHERI/glib/tree/glib_2.76.4_cheri based on https://github.com/arichardson/glib/tree/main-with-cheri-fixes, https://gitlab.gnome.org/arichardson/glib/-/tree/main-with-cheri-fixes, or yet something else? It would be good to discuss with @arichardson what's the CHERIfied upstream for glib.

@gcjenkinson Also, have you tested your branch with aarch64, not only aarch64c?

@markjdb While I'm fine with adding a patch to unblock some CheriABI packages, I wouldn't want to add it if it would break some aarch64 packages. If we're not sure about this, we should test this patch with aarch64 Chromiu, Ghidra, CHERI QEMU that we find crucial -- I can do that. If turns out some aarch64 programs are broken with it, and we don't have time to fix glib for them, we can apply the glib patch only when building for purecap.

**Comment by markjdb** [comment]
> @markjdb While I'm fine with adding a patch to unblock some CheriABI packages, I wouldn't want to add it if it would break some aarch64 packages. If we're not sure about this, we should test this patch with aarch64 Chromium, Ghidra, CHERI QEMU that we find crucial -- I can do that. If turns out some aarch64 programs are broken with it, and we don't have time to fix glib for them, we can apply the glib patch only when building for purecap.

I'm not really set up to test, so if you could I would really appreciate it. Note that setting `-DG_ENABLE_EXPERIMENTAL_ABI_COMPILATION` has no runtime effect, it just disables some compile-time assertions.

**Comment by arichardson** [comment]
> @gcjenkinson Is https://github.com/CTSRD-CHERI/glib/tree/glib_2.76.4_cheri based on https://github.com/arichardson/glib/tree/main-with-cheri-fixes, https://gitlab.gnome.org/arichardson/glib/-/tree/main-with-cheri-fixes, or yet something else? It would be good to discuss with @arichardson what's the CHERIfied upstream for glib.
> 
> @gcjenkinson Also, have you tested your branch with aarch64, not only aarch64c?
> 
> @markjdb While I'm fine with adding a patch to unblock some CheriABI packages, I wouldn't want to add it if it would break some aarch64 packages. If we're not sure about this, we should test this patch with aarch64 Chromium, Ghidra, CHERI QEMU that we find crucial -- I can do that. If turns out some aarch64 programs are broken with it, and we don't have time to fix glib for them, we can apply the glib patch only when building for purecap.

I managed to get most of the glib changes merged upstream, the outstanding diff is in https://gitlab.gnome.org/arichardson/glib/-/commits/main-with-cheri-fixes/?ref_type=heads

**Comment by kwitaszczyk** [comment]
@arichardson I have noticed that some commits from https://github.com/CTSRD-CHERI/glib/pull/1/commits were dropped, e.g. https://github.com/CTSRD-CHERI/glib/pull/1/commits/ea5345e3587805d4a3d3bddbb7d6abc62fa38437. Is it fine to drop them, or should they be re-added?

File: devel/glib20/Makefile [hunk 1]
```diff
@@ -34,6 +32,8 @@ MESON_ARGS=	-Db_lundef=false \
 		-Druntime_dir=/var/run \
 		-Dselinux=disabled \
 		-Dxattr=false
+
+CFLAGS+=	-DG_ENABLE_EXPERIMENTAL_ABI_COMPILATION
```
Comment (kwitaszczyk) [inline devel/glib20/Makefile:36]: It looks like this is still required with the GLib branch from https://github.com/CTSRD-CHERI/glib/pull/2. However, we do not need this for all architectures.
```suggestion
CFLAGS_aarch64c+=	-DG_ENABLE_EXPERIMENTAL_ABI_COMPILATION
```

Comment (markjdb) [reply devel/glib20/Makefile]: Fixed in the latest push.

**Comment by kwitaszczyk** [comment]
@markjdb Would you be able to test your use case of libslirp with devel/glib20 updated to https://github.com/CTSRD-CHERI/glib/tree/2.76.4-cheriabi-cherrypicks instead? I want to unblock things and if we can confirm that everything works fine in your case, I'd be tempted to go ahead with https://github.com/CTSRD-CHERI/glib/pull/2 and this PR.

**Comment by markjdb** [comment]
> @markjdb Would you be able to test your use case of libslirp with devel/glib20 updated to https://github.com/CTSRD-CHERI/glib/tree/2.76.4-cheriabi-cherrypicks instead? I want to unblock things and if we can confirm that everything works fine in your case, I'd be tempted to go ahead with [CTSRD-CHERI/glib#2](https://github.com/CTSRD-CHERI/glib/pull/2) and this PR.

Sure, I will test it this afternoon.

**Comment by markjdb** [comment]
> @markjdb Would you be able to test your use case of libslirp with devel/glib20 updated to https://github.com/CTSRD-CHERI/glib/tree/2.76.4-cheriabi-cherrypicks instead? I want to unblock things and if we can confirm that everything works fine in your case, I'd be tempted to go ahead with [CTSRD-CHERI/glib#2](https://github.com/CTSRD-CHERI/glib/pull/2) and this PR.

Sorry for the delay. The latest push regenerated the patch as you requested, and I verified that bhyve+slirp is still happy.

**Comment by kwitaszczyk** [comment]
> Sorry for the delay. The latest push regenerated the patch as you requested, and I verified that bhyve+slirp is still happy.

No worries. Thanks! I've merged the PR into https://github.com/CTSRD-CHERI/glib/commits/2.76.4-cheriabi/ . Could you update the commit message to include the correct URL and hash? The hash is different in the target branch for some reason even though there are no differences in the code and I didn't modify commit messages...

**Comment by markjdb** [comment]
> > Sorry for the delay. The latest push regenerated the patch as you requested, and I verified that bhyve+slirp is still happy.
> 
> No worries. Thanks! I've merged the PR into https://github.com/CTSRD-CHERI/glib/commits/2.76.4-cheriabi/ . Could you update the commit message to include the correct URL and hash? The hash is different in the target branch for some reason even though there are no differences in the code and I didn't modify commit messages...

Done. I verified that the patch generated from `2.76.4-cheriabi` is identical.

**Comment by kwitaszczyk** [comment]
Thank you!

FYI I've built hybrid ABI devel/glib with this PR and tested Chromium, Ghidra and QEMU with it. They seem to work fine.

**Review by kwitaszczyk — APPROVED** [review]

**Pull request merged**